src/swtpm/swtpm_nvstore_linear_file.c

```diff
@@ -0,0 +1,274 @@
+        goto fail;
+    }
+
+    if (st.st_size >= (uint32_t)sizeof(struct nvram_linear_hdr)) {
+        /* valid regular file-ish */
+        mmap_state.size = st.st_size;
+        mmap_state.can_truncate = true;
+    } else if (S_ISREG(st.st_mode)) {
+        /* too small, and regular file - make room for at least the header */
+        if (ftruncate(mmap_state.fd, sizeof(struct nvram_linear_hdr))) {
+            logprintf(STDERR_FILENO,
+                      "SWTPM_NVRAM_LinearFile_Mmap: Could not ftruncate file "
+                      "(too small): %s\n",
+                      strerror(errno));
+            rc = TPM_FAIL;
+            goto fail;
+        }
+
+        if (fstat(mmap_state.fd, &st)) {
+            logprintf(STDERR_FILENO,
+                      "SWTPM_NVRAM_LinearFile_Mmap: Could not stat file (after "
+                      "truncate): %s\n",
+                      strerror(errno));
+            rc = TPM_FAIL;
+            goto fail;
+        }
+
+        mmap_state.size = st.st_size;
+        mmap_state.can_truncate = true;
+    } else if (S_ISBLK(st.st_mode)) {
+        /* valid block device, can't resize, but can use as is */
+        if (ioctl(mmap_state.fd, BLKGETSIZE64, &bd_size)) {
+            logprintf(STDERR_FILENO,
+                      "SWTPM_NVRAM_LinearFile_Mmap: Could not get block device "
+                      "size): %s\n",
+                      strerror(errno));
+            rc = TPM_FAIL;
+            goto fail;
+        }
+
+        mmap_state.size = bd_size;
+        mmap_state.can_truncate = false;
+
+        if (mmap_state.size < (uint32_t)sizeof(struct nvram_linear_hdr)) {
+            logprintf(STDERR_FILENO, "SWTPM_NVRAM_LinearFile_Mmap: block device"
+                                     " too small, cannot resize\n");
+            rc = TPM_FAIL;
+            goto fail;
+        }
+    } else {
+        logprintf(STDERR_FILENO, "SWTPM_NVRAM_LinearFile_Mmap: invalid stat\n");
+        rc = TPM_FAIL;
+        goto fail;
+    }
+
+    mmap_state.ptr = mmap(NULL, mmap_state.size, PROT_READ | PROT_WRITE,
+                          MAP_SHARED, mmap_state.fd, 0);
+    if (mmap_state.ptr == MAP_FAILED) {
+        logprintf(STDERR_FILENO,
+                  "SWTPM_NVRAM_LinearFile_Mmap: Could not mmap file: %s\n",
+                  strerror(errno));
+        rc = TPM_FAIL;
+        goto fail;
+    }
+
+    return rc;
+
+fail:
+    close(mmap_state.fd);
+    return rc;
+}
+
+/*
+    Strip leading "file://" from uri if given.
+*/
+static const char*
+SWTPM_NVRAM_LinearFile_UriToPath(const char* uri)
+{
+    const char *path = uri;
+    if (strncmp(uri, "file://", 7) == 0) {
+        path += 7;
+    }
+    return path;
+}
+
+static TPM_RESULT
+SWTPM_NVRAM_LinearFile_Open(const char* uri,
+                            unsigned char **data,
+                            uint32_t *length)
+{
+    TPM_RESULT rc = 0;
+    const char *path = SWTPM_NVRAM_LinearFile_UriToPath(uri);
+
+    if (mmap_state.fd) {
```

Comment:
fd = 0 being a valid fd: mmap_state.fd >= 0  ?

Reply:
This is correct, but the resolution would be wrong (since fd is initialized to 0, but nothing is mapped at that point). I added a simple boolean "mapped" field to the state to encode this instead of relying on some special value integer fd, I think that's nicer.